orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java

```diff
@@ -72,6 +74,63 @@ private void replaceStages(PipelineTemplate pipelineTemplate) {
     }
   }
 
+  private void expandStagePartials(PipelineTemplate pipelineTemplate) {
+    List<StageDefinition> templateStages = pipelineTemplate.getStages();
+
+    List<StageDefinition> addStages = new ArrayList<>();
+    templateStages.stream().filter(StageDefinition::isPartialType).forEach(partialPlaceholder -> {
+      PartialDefinition partial = pipelineTemplate.getPartials().stream()
+        .filter(p -> p.getRenderedPartials().containsKey(partialPlaceholder.getId()))
+        .findFirst()
+        .orElseThrow(() -> new IllegalTemplateConfigurationException(String.format("Could not find rendered partial: %s", partialPlaceholder.getId())));
+
+      List<StageDefinition> stages = partial.getRenderedPartials().get(partialPlaceholder.getId());
+
+      // Create a graph so we can find the leaf nodes
+      createGraph(stages);
+
+      // get leaf nodes of the partial
+      Map<String, Integer> outOrder = new HashMap<>();
+      Set<StageDefinition> sorted = new LinkedHashSet<>();
+      Map<String, Status> state = new HashMap<>();
+
+      Map<String, StageDefinition> graph = stages
+        .stream()
+        .collect(
+          Collectors.toMap(StageDefinition::getId, i -> i)
+        );
+
+      graph.keySet().forEach(k -> outOrder.put(k, 0));
+      graph.forEach((k, v) -> dfs(k, sorted, state, graph, outOrder));
+      Set<String> leafNodes = sorted
+        .stream()
+        .filter(i -> outOrder.get(i.getId()) == 0)
+        .map(StageDefinition::getId)
+        .collect(Collectors.toSet());
+
+      // assign root nodes to placeholder's dependsOn value
+      stages.stream()
```

Comment:
Oh I see what you are doing. Just doesn't look obvious at first glance. Maybe an opportunity to make it readable.
IIRC, you are extracting all partial stages, then find where they fit in in the overall stage graph ?